compiler/rustc_span/src/span_encoding.rs

```diff
@@ -401,6 +401,63 @@ impl Span {
             }),
         }
     }
+
+    // For optimization we are interested in cases in which the parent is inline and the parent
+    // update doesn't change format. All non-inline or format changing scenarios require accessing
+    // interner and can fall back to `Span::new`.
+    #[inline]
+    pub fn with_parent(self, parent: Option<LocalDefId>) -> Span {
+        let data = match_span_kind! {
+            self,
+            InlineCtxt(span) => {
+                // Only if the new parent is `None` the format will be preserved.
+                if parent.is_none() {
+                    return self;
+                }
+                span.data()
+            },
+            InlineParent(span) => {
+                let old_parent = LocalDefId { local_def_index: DefIndex::from_u32(span.parent as u32) };
+                // FIXME: Is this tracking necessary?
+                (*SPAN_TRACK)(old_parent);
+                // Any `Some` small new parent will preserve the format.
+                // `None` may change the format to `InlineCtxt` if ctxt and len are small enough.
+                if let Some(parent) = parent
+                    && let parent32 = parent.local_def_index.as_u32()
+                    && parent32 <= MAX_CTXT {
+                    return InlineParent::span(span.lo, span.len_with_tag, parent32 as u16);
+                }
+                span.data()
+            },
+            PartiallyInterned(span) => span.data(),
+            Interned(span) => span.data(),
+        };
+
+        // We could not keep the span in the same inline format, fall back to the complete logic.
+        if let Some(old_parent) = data.parent {
+            // FIXME: Is this tracking necessary?
+            (*SPAN_TRACK)(old_parent);
```

Comment:
I've reproduced the pre-existing logic precisely and it tracks the old span.
It's pretty counter intuitive, but it may indeed be correct? We are effectively taking the old span's absolute position (which needs tracking) and produce a new span with a new parent out of it.
Similarly to https://github.com/rust-lang/rust/pull/126544/files#r1641782800 I'm also not sure.

Reply:
Yes, we should track the old span. When the old span has a parent, it encodes the absolute lo/hi byte positions, but incr. comp. is convinced it only encodes positions relative to the parent. If we output a span with no parent, incr. comp. will see those absolute positions appear out of thin air.
So we need to tell incr. comp. that we are accessing the absolute positions with `SPAN_TRACK`. The only case when it's not required is when `parent == old_parent`.